transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -286,23 +329,21 @@ protected void run() {
                     default:
                         // fallthrough
                 }
-
+                lastBlockingAmount.lazySet(System.nanoTime() - runStartTime);
                 final int ioRatio = this.ioRatio;
+                final long ioStartTime = System.nanoTime();
+                lastProcessedChannels.lazySet(tryProcessReady(events, strategy));
+                final long ioEndTime = System.nanoTime();
+                final long ioTime = ioEndTime - ioStartTime;
                 if (ioRatio == 100) {
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
                     runAllTasks();
                 } else {
-                    final long ioStartTime = System.nanoTime();
-
-                    if (strategy > 0) {
-                        processReady(events, strategy);
-                    }
-
-                    final long ioTime = System.nanoTime() - ioStartTime;
                     runAllTasks(ioTime * (100 - ioRatio) / ioRatio);
                 }
+                final long taskTime = System.nanoTime() - ioTime;
+                lastIoProcessingTime.lazySet(ioTime * 1000);
```

Comment:
I've done it both ways. Push is nicer in many regards, but tends to produce a larger API and that can cause waste to populate objects to pass as arguments that aren't used. It's also sometimes hard to do push as efficiently when it is on. For example, here it was pretty easy to have a local variable for `ioTime`, but how that would be done with the `MetricsTracker` is less obvious. The same can be true when extending it over time. You're more dependent on making each part of the metric its own method call, but then the tracker may have to re-combine that information. If you expect multiple consumers, it also becomes heavier weight.

All that is to say that push takes more work, but its benefits can make that worthwhile. That's why we'll be doing push in gRPC, for example. But I couldn't fault someone for doing poll-based.